jni/VuforiaNative.cpp

```diff
@@ -0,0 +1,607 @@
+/* VuforiaNative - AR4AndroidApplication Example
+ *
+ * Example Chapter 5
+ * accompanying the book
+ * "Augmented Reality for Android Application Development", Packt Publishing, 2013.
+ *
+ * Copyright � 2013 Jens Grubert, Raphael Grasset / Packt Publishing.
+ *
+ * This code is the proprietary information of Qualcomm Connected Experiences, Inc.
+ * Any use of this code is subject to the terms of the License Agreement for Vuforia Software Development Kit
+ * available on the Vuforia developer website.
+ *
+ * https://developer.vuforia.com
+ *
+ * This example was built from the ImageTarget example accompanying the Vuforia SDK
+ * https://developer.vuforia.com/resources/sample-apps/image-targets-sample-app
+ *
+ * This class is based on the ImageTarget.cpp from the Vuforia ImageTarget example
+ */
+
+#include <jni.h>
+#include <android/log.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include <math.h>
+
+#include <GLES2/gl2.h>
+#include <GLES2/gl2ext.h>
+
+#include <QCAR/QCAR.h>
+#include <QCAR/CameraDevice.h>
+#include <QCAR/Renderer.h>
+#include <QCAR/VideoBackgroundConfig.h>
+#include <QCAR/Trackable.h>
+#include <QCAR/TrackableResult.h>
+#include <QCAR/Tool.h>
+#include <QCAR/Tracker.h>
+#include <QCAR/TrackerManager.h>
+#include <QCAR/ObjectTracker.h>
+#include <QCAR/CameraCalibration.h>
+#include <QCAR/UpdateCallback.h>
+#include <QCAR/DataSet.h>
+#include <QCAR/Image.h>
+
+
+#include "MathUtils.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+// Screen dimensions:
+unsigned int screenWidth        = 0;
+unsigned int screenHeight       = 0;
+
+// Indicates whether screen is in portrait (true) or landscape (false) mode
+bool isActivityInPortraitMode   = false;
+
+// The projection matrix used for rendering virtual objects:
+QCAR::Matrix44F projectionMatrix;
+
+// Constants:
+static const float kObjectScale = 3.f;
+
+QCAR::DataSet* dataSetStonesAndChips    = 0;
+
+bool switchDataSetAsap          = false;
+
+//global variables
+JavaVM* javaVM = 0;
+jobject activityObj = 0;
+
+JNIEXPORT jint JNICALL
+JNI_OnLoad(JavaVM* vm,  void* reserved) {
+    LOG("JNI_OnLoad");
+    javaVM = vm;
+    return JNI_VERSION_1_4;
+}
+
+// Object to receive update callbacks from QCAR SDK
+//1) The QCAR_onUpdate method runs in a separate thread from the renderer, so OpenGL calls will not work.
+//2) The State object received in the QCAR_onUpdate method is only valid for the scope of the method.
+class AR4AndroidApplication_UpdateCallback : public QCAR::UpdateCallback
+{   
```

Comment:
This is the function to query the camera image and transfer it to the Java layer